src/key_scan.rs

```diff
@@ -234,32 +332,57 @@ impl<L: Ledger> BackgroundKeyScan<L> {
                 );
 
                 for txn in block.txns() {
-                    // Remove any records that were spent by this transaction.
-                    for n in txn.input_nullifiers() {
-                        if let Some((_, uid)) = self.records.remove(&n) {
-                            // If we removed a record that we had already discovered, prune it's
-                            // path from the Merkle tree.
-                            self.records_mt.forget(uid);
+                    if let KeyType::Sending(_) = self.key.clone().key_type() {
+                        // Remove any records that were spent by this transaction.
+                        for n in txn.input_nullifiers() {
+                            if let Some((_, uid)) = self.records.remove(&n) {
+                                // If we removed a record that we had already discovered, prune it's
+                                // path from the Merkle tree.
+                                self.records_mt.forget(uid);
+                            }
                         }
                     }
 
                     if let Some(records) = txn.output_openings() {
-                        // If the transaction exposes its records, add the records themselves if
-                        // they belong to us; forget their Merkle paths if they do not.
+                        // If the transaction exposes its records, add the records themselves if they
+                        // are viewable or freezable by, or belong to us; forget their Merkle paths if
+                        // they do not.
                         let mut received_records = vec![];
                         for record in records {
-                            let comm = RecordCommitment::from(&record);
-                            if record.pub_key == self.key.pub_key() {
-                                received_records.push(record.clone());
-                                let nullifier = self.key.nullify(
-                                    record.asset_def.policy_ref().freezer_pub_key(),
-                                    uid,
-                                    &comm,
-                                );
-                                // If the record belongs to us, add it to our records.
-                                self.records.insert(nullifier, (record, uid));
-                            } else {
-                                self.records_mt.forget(uid);
+                            match self.key.clone().key_type() {
+                                KeyType::Viewing(viewing_key) => {
+                                    if record.asset_def.policy_ref().viewer_pub_key()
+                                        == &viewing_key.pub_key()
+                                    {
+                                        received_records.push(record.clone());
```

Comment:
I think we also need to push to `self.records` here

Reply:
Yes! As I mentioned in another comment, I think `self.records` should be `HashSet` for viewing and freezing, then we can insert `record` to `self.records` without its nullifier.